File: packages/sdk/src/resources/pointOfSale.ts
import type { FluxisClient } from '../client.js';
import type {
  CreateNotificationSettingsRequest,
  CreateNotificationSettingsResponse,
  CreatePaymentRequestCheckoutRequest,
  CreatePaymentRequestRequest,
  CreatePointOfSaleRequest,
  NotificationSettings,
  PaymentRequestCheckoutResponse,
  PaymentRequestResponse,
  PointOfSale,
  UpdateNotificationSettingsRequest,
  UpdateNotificationSettingsResponse,
  UpdatePointOfSaleRequest,
} from '../types/pointOfSale.js';

export class PointOfSaleResource {
  constructor(private readonly client: FluxisClient) {}

  async list(): Promise<PointOfSale[]> {
    return this.client.request<PointOfSale[]>('GET', '/pos');
  }

  async get(posId: string): Promise<PointOfSale> {
    return this.client.request<PointOfSale>('GET', `/pos/${posId}`);
  }

  async create(data: CreatePointOfSaleRequest): Promise<PointOfSale> {
    return this.client.request<PointOfSale>('POST', '/pos', data);
  }

  async update(posId: string, data: UpdatePointOfSaleRequest): Promise<PointOfSale> {
    return this.client.request<PointOfSale>('PUT', `/pos/${posId}`, data);
  }

  async getNotifications(posId: string): Promise<NotificationSettings> {
    return this.client.request<NotificationSettings>('GET', `/pos/${posId}/notifications`);
  }

  async createNotifications(posId: string, data: CreateNotificationSettingsRequest): Promise<CreateNotificationSettingsResponse> {
    return this.client.request<CreateNotificationSettingsResponse>('POST', `/pos/${posId}/notifications`, data);
  }

  async updateNotifications(posId: string, data: UpdateNotificationSettingsRequest): Promise<UpdateNotificationSettingsResponse> {
    return this.client.request<UpdateNotificationSettingsResponse>('PUT', `/pos/${posId}/notifications`, data);
  }

  async createPaymentRequest(posId: string, data: CreatePaymentRequestRequest): Promise<PaymentRequestResponse> {
    return this.client.request<PaymentRequestResponse>('POST', `/pos/${posId}/payment-request`, data);
  }

  async getPaymentRequest(posId: string, paymentRequestId: string): Promise<PaymentRequestResponse> {
    return this.client.request<PaymentRequestResponse>('GET', `/pos/${posId}/payment-request/${paymentRequestId}`);
  }

  async createPaymentRequestCheckout(posId: string, data: CreatePaymentRequestCheckoutRequest): Promise<PaymentRequestCheckoutResponse> {
    return this.client.request<PaymentRequestCheckoutResponse>('POST', `/pos/${posId}/payment-request-checkout`, data);
  }
}

File: packages/backend/sdk-go/fluxis/point_of_sale.go
package fluxis

import (
	"context"
	"fmt"
)

const (
	defaultPOSPage     = 1
	defaultPOSPageSize = 50
)

// PointOfSaleService provides point of sale API methods.
type PointOfSaleService struct {
	client *Client
}

// List returns a paginated list of point of sale endpoints.
func (s *PointOfSaleService) List(ctx context.Context, opts *ListPointOfSaleOptions) (*ListPointOfSaleResponse, error) {
	params := map[string]string{
		"page":      intQuery(0, defaultPOSPage),
		"page_size": intQuery(0, defaultPOSPageSize),
	}
	if opts != nil {
		params["page"] = intQuery(opts.Page, defaultPOSPage)
		params["page_size"] = intQuery(opts.PageSize, defaultPOSPageSize)
		if opts.AccountID != "" {
			params["accountID"] = opts.AccountID
		}
	}

	raw, _, err := s.client.doRequest(ctx, "GET", "/pos", nil, buildQuery(params), true)
	if err != nil {
		return nil, err
	}
	resp, err := decodeData[ListPointOfSaleResponse](raw)
	if err != nil {
		return nil, err
	}
	return &resp, nil
}

// Get returns a point of sale by ID.
func (s *PointOfSaleService) Get(ctx context.Context, posID string) (*PointOfSale, error) {
	raw, _, err := s.client.doRequest(ctx, "GET", posPath(posID), nil, nil, true)
	if err != nil {
		return nil, err
	}
	pos, err := decodeData[PointOfSale](raw)
	if err != nil {
		return nil, err
	}
	return &pos, nil
}

// Create creates a new point of sale.
func (s *PointOfSaleService) Create(ctx context.Context, req *CreatePointOfSaleRequest) (*PointOfSale, error) {
	raw, _, err := s.client.doRequest(ctx, "POST", "/pos", req, nil, true)
	if err != nil {
		return nil, err
	}
	pos, err := decodeData[PointOfSale](raw)
	if err != nil {
		return nil, err
	}
	return &pos, nil
}

// Update updates a point of sale.
func (s *PointOfSaleService) Update(ctx context.Context, posID string, req *UpdatePointOfSaleRequest) (*PointOfSale, error) {
	raw, _, err := s.client.doRequest(ctx, "PUT", posPath(posID), req, nil, true)
	if err != nil {
		return nil, err
	}
	pos, err := decodeData[PointOfSale](raw)
	if err != nil {
		return nil, err
	}
	return &pos, nil
}

// Delete deletes a point of sale.
func (s *PointOfSaleService) Delete(ctx context.Context, posID string) error {
	_, _, err := s.client.doRequest(ctx, "DELETE", posPath(posID), nil, nil, true)
	return err
}

// GetPaymentIntention returns the active payment intention for an open PoS.
func (s *PointOfSaleService) GetPaymentIntention(ctx context.Context, posID string) (*PaymentIntentionResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "GET", posPath(posID)+"/payment-intention", nil, nil, true)
	if err != nil {
		return nil, err
	}
	intention, err := decodeData[PaymentIntentionResponse](raw)
	if err != nil {
		return nil, err
	}
	return &intention, nil
}

// CreatePaymentIntention creates a payment intention for an open PoS.
func (s *PointOfSaleService) CreatePaymentIntention(ctx context.Context, posID string, req *CreatePaymentIntentionRequest) (*CreatePaymentIntentionResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "POST", posPath(posID)+"/payment-intention", req, nil, true)
	if err != nil {
		return nil, err
	}
	intention, err := decodeData[CreatePaymentIntentionResponse](raw)
	if err != nil {
		return nil, err
	}
	return &intention, nil
}

// ClosePaymentIntention closes the active payment intention.
func (s *PointOfSaleService) ClosePaymentIntention(ctx context.Context, posID string) error {
	_, _, err := s.client.doRequest(ctx, "POST", posPath(posID)+"/payment-intention/close", nil, nil, true)
	return err
}

// GetQR returns the QR code data for a PoS.
func (s *PointOfSaleService) GetQR(ctx context.Context, posID string) (*GetQRResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "GET", posPath(posID)+"/qr", nil, nil, true)
	if err != nil {
		return nil, err
	}
	qr, err := decodeData[GetQRResponse](raw)
	if err != nil {
		return nil, err
	}
	return &qr, nil
}

// CreatePaymentRequest creates a crypto payment request and returns a NASPIP token.
func (s *PointOfSaleService) CreatePaymentRequest(ctx context.Context, posID string, req *CreatePaymentRequestRequest) (*PaymentRequestResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "POST", posPath(posID)+"/payment-request", req, nil, true)
	if err != nil {
		return nil, err
	}
	payment, err := decodeData[PaymentRequestResponse](raw)
	if err != nil {
		return nil, err
	}
	return &payment, nil
}

// GetPaymentRequest returns a payment request by ID.
func (s *PointOfSaleService) GetPaymentRequest(ctx context.Context, posID, paymentRequestID string) (*PaymentRequestResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "GET", posPath(posID)+"/payment-request/"+paymentRequestID, nil, nil, true)
	if err != nil {
		return nil, err
	}
	payment, err := decodeData[PaymentRequestResponse](raw)
	if err != nil {
		return nil, err
	}
	return &payment, nil
}

// CreatePaymentRequestCheckout creates a fiat checkout payment request.
func (s *PointOfSaleService) CreatePaymentRequestCheckout(ctx context.Context, posID string, req *CreatePaymentRequestCheckoutRequest) (*PaymentRequestCheckoutResponse, error) {
	raw, _, err := s.client.doRequest(ctx, "POST", posPath(posID)+"/payment-request-checkout", req, nil, true)
	if err != nil {
		return nil, err
	}
	payment, err := decodeData[PaymentRequestCheckoutResponse](raw)
	if err != nil {
		return nil, err
	}
	return &payment, nil
}

func posPath(posID string) string {
	return fmt.Sprintf("/pos/%s", posID)
}
